dns/miekgdns2/resolver.go

```diff
@@ -0,0 +1,356 @@
+	"sync"
+	"time"
+
+	"github.com/miekg/dns"
+
+	"github.com/grafana/dskit/multierror"
+)
+
+const (
+	DefaultResolvConfPath  = "/etc/resolv.conf"
+	defaultMaxConnsPerHost = 2
+)
+
+type Client interface {
+	Exchange(ctx context.Context, msg *dns.Msg, server string) (*dns.Msg, time.Duration, error)
+	Clean(known []string)
+}
+
+// Resolver is a DNS service discovery backend that retries on errors, only uses TCP,
+// and pools connections to nameservers to increase reliability.
+//
+// This backend:
+// * Does _not_ use search domains, all names are assumed to be fully qualified
+// * Only uses TCP connections to the nameservers
+// * Keeps several connections to each nameserver open
+// * Reads resolv.conf on every query
+// * Closes connections to unknown nameservers on every query
+type Resolver struct {
+	confPath string
+	client   Client
+}
+
+// NewResolver creates a new Resolver that uses the provided resolv.conf configuration
+// to perform DNS queries.
+func NewResolver(resolvConf string) *Resolver {
+	return NewResolverWithClient(resolvConf, NewPoolingClient(defaultMaxConnsPerHost))
+}
+
+// NewResolverWithClient creates a new Resolver that uses the provided resolv.conf configuration
+// and Client implementation to perform DNS queries.
+func NewResolverWithClient(resolvConf string, client Client) *Resolver {
+	return &Resolver{
+		confPath: resolvConf,
+		client:   client,
+	}
+}
+
+func (r *Resolver) IsNotFound(error) bool {
+	// We don't return an error when there are no hosts found so this is
+	// always false. Instead, we return the empty DNS response with the
+	// appropriate return code set.
+	return false
+}
+
+func (r *Resolver) LookupSRV(ctx context.Context, service, proto, name string) (cname string, addrs []*net.SRV, err error) {
+	servers, conf, err := r.loadConfiguration()
+	if err != nil {
+		return "", nil, err
+	}
+
+	// Close connections to nameservers no longer configured in resolv.conf
+	r.client.Clean(servers)
+	return r.lookupSRV(ctx, servers, conf, service, proto, name)
+}
+
+func (r *Resolver) lookupSRV(ctx context.Context, servers []string, conf *dns.ClientConfig, service, proto, name string) (cname string, addrs []*net.SRV, err error) {
+	var target string
+	if service == "" && proto == "" {
+		target = name
+	} else {
+		target = "_" + service + "._" + proto + "." + name
+	}
+
+	response, err := r.query(ctx, servers, conf.Attempts, target, dns.Type(dns.TypeSRV))
+	if err != nil {
+		return "", nil, err
+	}
+
+	for _, record := range response.Answer {
+		switch addr := record.(type) {
+		case *dns.SRV:
+			addrs = append(addrs, &net.SRV{
+				Weight:   addr.Weight,
+				Target:   addr.Target,
+				Priority: addr.Priority,
+				Port:     addr.Port,
+			})
+		default:
+			return "", nil, fmt.Errorf("invalid SRV response record %s", record)
+		}
+	}
+
+	return "", addrs, err
+}
+
+func (r *Resolver) LookupIPAddr(ctx context.Context, host string) ([]net.IPAddr, error) {
+	servers, conf, err := r.loadConfiguration()
```

Comment:
Should we cache the config? Seems expensive to parse it for every DNS query, although maybe those are infrequent enough that it doesn't matter.

Reply:
This is what the existing resolver did but it does seem expensive. I don't think it'll be an issue but I can look into caching it for a short amount of time.

Reply:
I've added some logic to reload the configuration and clean up defunct connections in a separate goroutine